cloud/cloud_env_impl.cc

```diff
@@ -1840,6 +1844,150 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
   return Status::OK();
 }
 
+Status CloudEnvImpl::RollCloudManifest(const std::string& local_dbname,
+                                       const std::string& new_cookie,
+                                       std::string* existing_cookie_out) {
+  assert(HasDestBucket());
+  std::string current_cookie = GetCookie();
+  if (current_cookie == new_cookie) {
+    return Status::InvalidArgument(
+        "Rolling cloud manifest with same cookie is illegal");
+  }
+
+  // Fetch cloud manifest to local
+  // TODO(wei): since CLOUDMANIFEST-cookie file is immutable (for the same
+  // cookie), we can avoid fetching from cloud if it exists locally
+  Status st = FetchCloudManifest(local_dbname);
+  if (!st.ok()) {
+    return st;
+  }
+
+  // Loading local cloud manifest file
+  std::unique_ptr<CloudManifest> cloud_manifest;
+  st = LoadLocalCloudManifest(local_dbname, GetBaseEnv(),
+                              GetCookie(), &cloud_manifest);
+  if (!st.ok()) {
+    return st;
+  }
+
+  // Fetch latest Manifest file corresponds to current cookie
+  // NOTE: the latest Manifest file could be different from local Manifest file.
+  // TODO(wei): should we try to avoid overwriting the local Manifest file (with same name/epoch)?
+  st = FetchCurrentManifest(local_dbname);
+  if (!st.ok()) {
+    return st;
+  }
+
+  auto old_epoch = cloud_manifest->GetCurrentEpoch();
+
+  // Get the max file number of latest Manifest file.
+  // NOTE: the max file number could be different from VersionSet::next_file_num
+  uint64_t max_file_num;
+  st = ManifestReader::GetMaxFileNumberFromManifest(
+      GetBaseEnv(), ManifestFileWithEpoch(local_dbname, old_epoch),
+      &max_file_num);
+  if (!st.ok()) return st;
+
+  // rolling epoch of the Manifest file
+  auto new_epoch = generateNewEpochId();
+  cloud_manifest->AddEpoch(max_file_num, new_epoch);
```

Comment:
@igorcanadi Can we assume that there is only one single thread calling `Roll/Switch CloudManifest` for one shard?

It can be an issue when it's called in multi-threads since `cloud_manifest` can point to different content here (compared to calling `GetCurrentEpoch` above).

Or we should move thread safety protection for cloud manifest into `CloudEnv` by protecting the pointer.

Reply:
Yes, we should add this to the contract of the API. It shouldn't be the problem to guarantee that on the caller.